language/movie_nl/main.py

```diff
@@ -0,0 +1,389 @@
+            raise StopIteration()
+
+        doc_id = os.path.basename(item)
+
+        with codecs.open(item, encoding='utf-8') as f:
+            try:
+                text = f.read()
+            except UnicodeDecodeError:
+                continue
+
+            yield Document(text, doc_id, item)
+
+
+def rank_entities(reader, sentiment=None, topn=None, reverse_bool=False):
+    """Rank the entities (actors) based on their sentiment
+    assigned from the movie."""
+
+    items = []
+    for item in reader:
+        json_item = json.loads(item)
+        sent = json_item.get('sentiment')
+        entity_item = (sent, json_item)
+
+        if sentiment:
+            if sentiment == 'pos' and sent > 0:
+                items.append(entity_item)
+            elif sentiment == 'neg' and sent < 0:
+                items.append(entity_item)
+        else:
+            items.append(entity_item)
+
+    items.sort(reverse=reverse_bool)
+    items = [json.dumps(item[1]) for item in items]
+
+    print('\n'.join(items[:topn]))
+
+
+def get_service():
+    """Build a client to the Google Cloud Natural Language API."""
+
+    credentials = GoogleCredentials.get_application_default()
+
+    return discovery.build('language', 'v1beta1',
+                           credentials=credentials)
+
+
+def analyze(input_dir, sentiment_writer, entity_writer, sample, log_file):
+    """Analyze the document for sentiment and entities"""
+
+    # Create logger settings
+    logging.basicConfig(filename=log_file, level=logging.DEBUG)
+
+    # Create a Google Service object
+    service = get_service()
+
+    reader = document_generator(input_dir, sample)
+
+    # Process the movie documents
+    process_movie_reviews(service, reader, sentiment_writer, entity_writer)
+
+    # close reader and writers
+    sentiment_writer.close()
+    entity_writer.close()
+    reader.close()
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(
+        description=__doc__,
+        formatter_class=argparse.RawDescriptionHelpFormatter)
+
+    subparsers = parser.add_subparsers(dest='command')
+
+    rank_parser = subparsers.add_parser('rank')
+
+    rank_parser.add_argument(
+        'entity_input', help='location of entity input',
+        type=argparse.FileType('r'))
+    rank_parser.add_argument(
+        '--sentiment', help='filter sentiment as "neg" or "pos"')
+    rank_parser.add_argument(
+        '--reverse', help='reverse the order of the items', type=bool,
+        default=False
+        )
+    rank_parser.add_argument(
+        '--sample', help='number of top items to process', type=int,
+        default=None
+        )
+
+    analyze_parser = subparsers.add_parser('analyze')
+
+    analyze_parser.add_argument(
+        '--inp', help='location of the input', required=True)
+    analyze_parser.add_argument(
+        '--sout', help='location of the sentiment output', required=True,
+        type=argparse.FileType('w'))
```

Comment:
Nice for who? The person reading this for the first time will likely have
to go to the argparse docs to figure out what exactly is going on here.

On Thu, Jul 21, 2016, 11:22 AM puneith <notifications@github.com> wrote:

> In language/movie_nl/main.py
> <https://github.com/GoogleCloudPlatform/python-docs-samples/pull/426#discussion_r71759502>
> :
>
> > +    rank_parser.add_argument(
> > +        '--reverse', help='reverse the order of the items', type=bool,
> > +        default=False
> > +        )
> > +    rank_parser.add_argument(
> > +        '--sample', help='number of top items to process', type=int,
> > +        default=None
> > +        )
> > +
> > +    analyze_parser = subparsers.add_parser('analyze')
> > +
> > +    analyze_parser.add_argument(
> > +        '--inp', help='location of the input', required=True)
> > +    analyze_parser.add_argument(
> > +        '--sout', help='location of the sentiment output', required=True,
> > +        type=argparse.FileType('w'))
>
> Actually it never ended with that conclusion. The debate was still on. To
> actually think of it this does abstract away the file opening which is nice
> :)
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/python-docs-samples/pull/426/files/850771d816fbe71389ad1e94447400b07069dd0e#r71759502>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAPUcz57ynb6YEis9PVXr7TxMHS5rPgiks5qX7j2gaJpZM4JRN78>
> .
>